mqttools/client.py

```diff
@@ -545,10 +545,12 @@ async def unsubscribe(self, topic):
             if reason != UnsubackReasonCode.SUCCESS:
                 raise UnsubscribeError(reason)
 
-    def publish(self, topic, message, retain=False):
+    def publish(self, topic, message, retain=False, properties=None):
```

Comment:
Hmmm, I've been thinking about if it's a good idea to expose the properties dictionary to the user. It's kind of an implementation detail she shouldn't know about. I know it's currently part of the `on_message()` method, which should be reconsidered.

I think the publish method should look something like
```python
def publish(self, message):
```
where `message` is something like
```python
class Message:

    def __init__(self, topic, message, retain=False, response_topic=None):
        self.topic = topic
        self.message = message
        self.retain = retain
        self.response_topic = response_topic
```
`on_message()` will be
```python
async def on_message(self, message):
```
and the messages queue will contain messages.

I think I can merge this PR with its current design, and redesign the API after that as it's a big change.

Reply:
You are right in keeping the user side constant.
I like that the queues then contain simply Messages instead of untyped tuples
Message can be quite concise using a dataclass:

```python
@dataclass
class Message:
    topic: str
    message: bytes
    retain: bool = False
    response_topic: str = None
```